File: Day-7/solution.cpp
#include <iostream>
#include <vector>
using namespace std;

int maxProfit(vector<int>& prices) {
    if (prices.size() < 2)
        return 0;

    int minPrice = prices[0];
    int maxProfit = 0;

    for (int i = 1; i < prices.size(); i++) {
        // update profit if selling today is better
        maxProfit = max(maxProfit, prices[i] - minPrice);

        // update minimum buying price
        minPrice = min(minPrice, prices[i]);
    }

    return maxProfit;
}

int main() {
    vector<int> prices = {7, 1, 5, 3, 6, 4};
    cout << maxProfit(prices);
    return 0;
}

File: Day-7/solution.py
from typing import List

def maxProfit(prices: List[int]) -> int:
    if len(prices) < 2:
        return 0

    min_price = prices[0]
    max_profit = 0

    for price in prices[1:]:
        # profit if we sell today
        max_profit = max(max_profit, price - min_price)

        # update minimum buying price
        min_price = min(min_price, price)

    return max_profit


# Example
prices = [7, 1, 5, 3, 6, 4]
print(maxProfit(prices))  # Output: 5
